### PR TITLE
Implement algorithms

### PR DESCRIPTION
**Pull request by pauljohanneskraft**
# Algorithms

## :recycle: Current situation & Problem
This PR adds algorithms to recommend medication changes based on vitals, contraindications and existing medications.

## :gear: Release Notes 
Added algorithms to recommend medication changes based on vitals, contraindications and existing medication.

## :books: Documentation
-

## :white_check_mark: Testing
-

### Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: functions/src/services/recommendation/recommenders/betaBlockersRecommender.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+      case ContraindicationCategory.none:
+        break
+    }
+
+    const medianSystolic = input.vitals.systolicBloodPressure.at(
+      input.vitals.systolicBloodPressure.length / 2,
+    )?.value
+    const medianHeartRate = input.vitals.heartRate.at(
+      input.vitals.heartRate.length / 2,
+    )?.value
```
Comment (PSchmiedmayer) [inline functions/src/services/recommendation/recommenders/betaBlockersRecommender.ts:69]: Not sure if I understand this code in a right way, but does this actually calculate the median of the values? I just seems to access the middle element?

Seems like this is also applying to other parts of the code.

File: functions/src/services/recommendation/recommenders/recommender.ts
Comment (PSchmiedmayer) [inline functions/src/services/recommendation/recommenders/recommender.ts:1]: Nice way to extract this. We might also move some of the median calculations and general values used across the functions into this type?

**Review by PSchmiedmayer — COMMENTED** [review]
Thank you for working on the PR @pauljohanneskraft!

I had a few first comments; it will be important that this code is especially well tested with examples and possible edge cases.

As noted in the meeting today, we should start to identify some common test cases and need to ensure that we then ask Alex and @vishnuravi for some more test cases and possible edge cases that we need to account for.

File: functions/src/services/recommendation/recommenders/recommender.ts
Comment (pauljohanneskraft) [reply functions/src/services/recommendation/recommenders/recommender.ts:1]: extracted them into functions in 'extensions/array.ts'

File: functions/src/services/recommendation/recommenders/betaBlockersRecommender.ts (hunk 1 above)
Comment (pauljohanneskraft) [reply functions/src/services/recommendation/recommenders/betaBlockersRecommender.ts]: used the extracted version and that one should work fine.

File: functions/src/services/recommendation/recommenders/diureticRecommender.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+  compute(input: RecommendationInput): MedicationRecommendation[] {
+    const currentMedication = this.findCurrentMedication(input.requests, [
+      MedicationClassReference.diuretics,
+    ])
+    if (!currentMedication) return []
+    return this.createRecommendation(
+      currentMedication,
+      undefined,
+      MedicationRecommendationCategory.noActionRequired,
```
Comment (PSchmiedmayer) [inline functions/src/services/recommendation/recommenders/diureticRecommender.ts:28]: Should probably be `personalTargetDoseReached`.

File: functions/src/services/recommendation/recommenders/diureticRecommender.ts
Comment (PSchmiedmayer) [inline functions/src/services/recommendation/recommenders/diureticRecommender.ts:1]: Would be great to document the reasoning for this recommender.

File: functions/src/services/recommendation/recommendationService.ts
Comment (PSchmiedmayer) [inline functions/src/services/recommendation/recommendationService.ts:1]: Would be great to document the meaning of the recommender outputs and the correlation to target icons beyond the Google Docs document in the README as part of this PR.

**Review by PSchmiedmayer — APPROVED** [review]
Great job with the PR and logic @pauljohanneskraft!

Took a look at the different algorithms and the remaining work is only fixing a few smaller things and enduring that we translate most elements from the Google Docs and reasoning into this PR, README, and Documentation to ensure that future maintainers can follow all decisions right within the repo and code.

File: functions/src/services/symptomScoreCalculator/defaultSymptomScoreCalculator.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+  question2 = '692bda7d-a616-43d1-8dc6-8291f6460ab2',
+  question3 = 'b1734b9e-1d16-4238-8556-5ae3fa0ba913',
+  question4 = '57f37fb3-a0ad-4b1f-844e-3f67d9b76946',
+  question5 = '396164df-d045-4c56-d710-513297bdc6f2',
+  question6 = '75e3f62e-e37d-48a2-f4d9-af2db8922da0',
+  question7 = 'fce3a16e-c6d8-4bac-8ab5-8f4aee4adc08',
+  question8a = '8b022e69-127d-4447-8190-39ac645e60e1',
+  question8b = '1eee7259-da1c-4cba-80a9-e67e684573a1',
+  question8c = '883a22a8-2f6e-4b41-84b7-0028ed543192',
+  question9 = '24108967-2ff3-40d0-c54f-a7b97bb84d05',
```
Comment (PSchmiedmayer) [inline functions/src/services/symptomScoreCalculator/defaultSymptomScoreCalculator.ts:30]: We should probably think about given the link ids human readable names in the FHIR questionnaire to make this mapping easier and also in the long-run easier to maintain here & in the questionnaire as things change. The UUIDs are really hard to identify and compare.

File: functions/src/tests/mocks/questionnaire.ts
Comment (PSchmiedmayer) [inline functions/src/tests/mocks/questionnaire.ts:1]: Wondering if we should rather use the questionnaire seeded here as a reference to also somehow test that the seeded questionnaire is correct and interpretable?

**Review by PSchmiedmayer — APPROVED** [review]
Thank you for working on all of this here in the PR! 🚀

Only had a few minor comments looking at the different elements in this PR with the symptom score calculations added 👍

File: functions/src/services/healthSummaryService.ts [hunk 4]
```diff
@@ -57,16 +69,35 @@ export class HealthSummaryService {
       medications: medications,
       vitals: {
         ...vitals,
-        dryWeight: user.content?.dryWeight?.value ?? vitals.dryWeight,
+        dryWeight:
```
[the remark this answers is not in the document]
Comment (pauljohanneskraft) [reply functions/src/services/healthSummaryService.ts]: They are being retrieved from the respective collections in service-separation

File: functions/src/services/recommendation/recommenders/betaBlockerRecommender.ts
[the remark this answers is not in the document]
Comment (pauljohanneskraft) [reply functions/src/services/recommendation/recommenders/betaBlockerRecommender.ts:1]: done for the algorithm diagrams - no explanation yet

File: functions/src/services/recommendation/recommendationService.ts
Comment (pauljohanneskraft) [reply functions/src/services/recommendation/recommendationService.ts:1]: done

File: functions/src/services/symptomScoreCalculator/defaultSymptomScoreCalculator.ts (hunk 3 above)
Comment (pauljohanneskraft) [reply functions/src/services/symptomScoreCalculator/defaultSymptomScoreCalculator.ts:30]: I have made this so that it may work with different questionnaires - not sure about changing the ids and how that would easily be doable with the tool you showed me - if there is an easy way, we can adapt it later

File: functions/src/tests/mocks/questionnaire.ts
Comment (pauljohanneskraft) [reply functions/src/tests/mocks/questionnaire.ts:1]: makes sense - this is the same as in questionnaires.json btw
